### ISSUE TITLE
Generate browsable command documentation

### ISSUE THREAD

**Issue by jasonmitchell** [jasonmitchell/sequin#13]
Might be useful to be able to automatically generate command documentation that can be browsed (like Swagger documentation) to make command discovery easier.

Possibly another piece of middleware.

**Comment by JoeStead** [comment]
I think JSON style documentation should be automagically generated (like what you already have), but I think it'd be good to have some extra metadata included, so default endpoint (is it URI/header based for example), and potentially a tool to generate sample requests, so it's _really_ like Swagger. The tool should obviously be separated from the "usage feed" JSON. A tool of this kind is potentially out of scope from this, but configuration options are not

**Comment by jasonmitchell** [comment]
Would it be possible to use swagger to generate browsable documentation?  This would cut a load of effort out.

The more documentation that can be generated for commands the better.

**Comment by JoeStead** [comment]
Based on the [Swagger Specs](http://swagger.io/specification/) you could build up some docs with relative ease. The description tag would almost definitely have to be done with XML comments, but the other stuff can almost definitely be inferred from the commands and a bit of reflection magic. It will get a bit messy with different resolution implementations (header vs URI, where URI is definitely simpler)

**Comment by JoeStead** [comment]
Swagger probably isn't the best idea for this. It isn't a REST API as such. 

Generating some documentation based on the commands should be quite simple, but it gets a little more interesting when you take into account validation. I've got a working sample of pulling out the rules for validators, so I could easily hook into the FluentValidation stuff and generate some docs from that - We'd probably want to provide an interface for other validation implementations to hook into. 

I'm thinking for now, just Markdown files?

**Comment by JoeStead** [comment]
I've created a bastardised solution with all of the relevant projects in, I'll try and summarise some of the changes I've made to get documentation **possible** for validation.

1. I've added a new project called Sequin.Validation to handle the common stuff (this could probably be pulled into Sequin and Sequin.Core respectively).
2. Added a "SequinValidationResult" which has two props `bool IsValid` and `object Result`
3. I've added an ISequinValidatorFactory which looks like this:

In summary we've got an interface to hook into to generate validator documentation for commands, but I'm not sure the approach is the best way. I toyed with the idea of having a `ISequinValidator` but I don't think that should bleed out into the users application, so I've gone with the SequinValidationResult / Func to solve it. [Branch can be found here](https://github.com/JoeStead/sequin/tree/feature-documentation). The interesting bits are the two validation projects.
